src/gbinder_reader.c

```diff
@@ -431,89 +430,94 @@ gbinder_reader_read_hidl_vec1(
+    return g_strdup(gbinder_reader_read_hidl_string_c(reader));
 }
 
 char**
 gbinder_reader_read_hidl_string_vec(
     GBinderReader* reader)
 {
-    GBinderBuffer* buf = gbinder_reader_read_buffer(reader);
+    GBinderIoBufferObject obj;
 
     /* First buffer contains hidl_vector */
-    if (buf && buf->size == sizeof(GBinderHidlVec)) {
-        GBinderHidlVec* vec = buf->data;
+    if (gbinder_reader_read_buffer_object(reader, &obj) &&
+        obj.data && obj.size == sizeof(GBinderHidlVec)) {
+        GBinderHidlVec* vec = obj.data;
         const guint n = vec->count;
         const void* next = vec->data.ptr;
 
-        gbinder_buffer_free(buf);
         if (!next && !n) {
-            char** out = g_new(char*, 1);
-
-            out[0] = NULL;
-            return out;
-        } else {
-            /* The second buffer (if any) contains n hidl_string's */
-            buf = gbinder_reader_read_buffer(reader);
-            if (buf && buf->data == next &&
-                buf->size == (sizeof(GBinderHidlString) * n)) {
-                const GBinderHidlString* strings = buf->data;
-                GBinderBuffer* sbuf;
-                GPtrArray* list = g_ptr_array_new();
-                guint i;
-
-                /* Now we expect n buffers containing the actual data */
-                for (i=0; i<n &&
-                    (sbuf = gbinder_reader_read_buffer(reader)); i++) {
-                    const GBinderHidlString* s = strings + i;
-                    if (sbuf->size == s->len + 1 &&
-                        sbuf->data == s->data.str &&
-                        s->data.str[s->len] == 0) {
-                        char* name = g_strdup(s->data.str);
-
-                        g_ptr_array_add(list, name);
-                        GVERBOSE_("%u. %s", i + 1, name);
-                        gbinder_buffer_free(sbuf);
-                    } else {
-                        GWARN("Unexpected hidl_string buffer %p/%u vs %p/%u",
-                            sbuf->data, (guint)sbuf->size, s->data.str, s->len);
-                        gbinder_buffer_free(sbuf);
-                        break;
-                    }
-                }
-
-                if (i == n) {
-                    gbinder_buffer_free(buf);
-                    g_ptr_array_add(list, NULL);
-                    return (char**)g_ptr_array_free(list, FALSE);
+            /* Should this be considered an error? */
+            return g_new0(char*, 1);
+        } else if (gbinder_reader_read_buffer_object(reader, &obj) &&
+                   /* The second buffer (if any) contains n hidl_string's */
+                   obj.parent_offset == GBINDER_HIDL_VEC_BUFFER_OFFSET &&
+                   obj.has_parent &&
+                   obj.data == next &&
+                   obj.size == (sizeof(GBinderHidlString) * n)) {
+            const GBinderHidlString* strings = obj.data;
+            GPtrArray* list = g_ptr_array_new();
```

Comment:
Since we already have n here, would it make sense to pre-allocate the array to n + 1 to avoid reallocs?

Reply:
Right. Done!